src/hotspot/share/utilities/lineReader.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.inline.hpp"
+#include "runtime/os.hpp"
+#include "utilities/lineReader.hpp"
+
+LineReader::LineReader(const char* filename) : _filename(filename), _stream(nullptr) {
+  // Use os::open() because neither fopen() nor os::fopen()
+  // can handle long path name on Windows. HMM, is this still valid today???
+  int fd = os::open(filename, O_RDONLY, S_IREAD);
+  if (fd != -1) {
+    // Obtain a File* from the file descriptor so that getc()
+    // can be used in get_line().
+    _stream = os::fdopen(fd, "r");
+    if (_stream == nullptr) {
+      ::close(fd);
+    } else {
+      // fd will be closed by fclose(_stream)
+    }
+  } else {
+    _stream = nullptr;
+  }
+
+  _buffer_length = 32;
+  _buffer = NEW_RESOURCE_ARRAY(char, _buffer_length);
+}
+
+LineReader::~LineReader() {
+  close(); // Just in case
+}
+
+void LineReader::close() {
+  if (_stream != nullptr) {
+    fclose(_stream);
+    _stream = nullptr;
+  }
+}
+
+// Returns nullptr if we have reached EOF.
+// \n is treated as the line separator.
+// All occurrences of \r are stripped.
+char* LineReader::get_line() {
+  if (_stream == nullptr) {
+    return nullptr;
+  }
+  size_t buffer_pos = 0;
+  int c;
+  while ((c = getc(_stream)) != EOF) {
```

Comment:
Lets not read individual characters. Lets use fgets or fread() or just plain read(). Preferably the first.

Reply:
I've reimplemented using fgets().

Reply:
So now this is not just a refactoring it is a re-implementation, so now I have to try and understand the reading logic and can't just treat it as known good code. :(

Reply:
I can revert to the getc version and implement the fgets version in a separate RFE. I agree that keeping the logic unchanged in refactoring will make things more manageable.

getc is buffered I/O, so there shouldn't be much difference in performance when compared to fgets. And the code affected by this PR doesn't care too much about I/O performance.

Reply:
By the way, I think I should take out the handling of _max_buffer_length, which I actually haven't implemented. I don't know what the failure mode should be when the line width is longer than the specified max, and I don't want to write any test case when the max is set to a small number like 256.

So let's keep the existing 2x expansion and crash when OOM (which is the same as the numerous sites that use GrowableArray).

Reply:
Okay, pity since that makes the code less reusable for cases I had in mind. For instance, reading and writing out gigantic mapping files. I recently had a customer with a 4TB heap and ZGC, where /proc/pid/maps alone was 20mio lines.